### DISCUSSION TITLE
Proxy Receive Notification

### DISCUSSION THREAD

**Discussion by csehatt741** [microsoft/vs-streamjsonrpc#821] [Q&A]
Following the [Dynamically generated proxies](https://github.com/microsoft/vs-streamjsonrpc/blob/main/doc/dynamicproxy.md) I'm still struggling with configuring the proxy to properly receive notifications. Though, the notification is received, but the configured `CustomEventArgs` is always `null`.

Proxy interface:
```csharp
public interface IProxy
{
    event EventHandler<CustomEventArgs> OnEvent;
}
```

Custom EventArgs:
```csharp
public class CustomEventArgs : EventArgs
{
    public object Data { get; set; }
    public string @Object { get; set;  }
    public string Type { get; set; }
}
```

JSON-RPC message:
```json
{
  "jsonrpc": "2.0",
  "method": "onEvent",
  "params": {
    "value": {
      "data": {
        "source": "681f1bc8-2d13-4189-a82a-2e2b92248a21_kurento.MediaPipeline/e983997e-ac19-4f4b-9575-3709af8c01be_kurento.PlayerEndpoint",
        "tags": [],
        "timestampMillis": "1441277150433",
        "type": "EndOfStream"
      },
      "object": "681f1bc8-2d13-4189-a82a-2e2b92248a21_kurento.MediaPipeline/e983997e-ac19-4f4b-9575-3709af8c01be_kurento.PlayerEndpoint",
      "type": "EndOfStream"
    }
  }
}
```

Proxy generation:
```csharp
var proxy = JsonRpc.Attach<IProxy>(messageHandler, proxyOptions);

proxy.OnEvent += (object? sender, CustomEventArgs args) => { };
```

**Comment by AArnott** [comment]
I wonder if the reason your `args` comes through as null is because the JSON-RPC message says the parameter name is `value`. StreamJsonRpc tries to match the parameter name of the method receiving the message to the argument name given in the message, and a mismatch may lead to it being dropped. Normally it would lead to the method not being invoked at all, but it may be special for events, I'm not sure.
Another possibility is that your `CustomEventArgs` class is defined without any attributes, so deserialization of the event argument doesn't actually match any of the properties on your class.

Finally, and a separate issue, if you're going to wire up events *after* using the static `Attach<T>` method, just be aware that events that come in *before* you add the event handler will be dropped. If you can't afford events to be dropped, you should add all handler _before_ `StartListening` is invoked. The `Attach<T>` methods call `StartListening` for you, so you would instead want to call the `JsonRpc` constructor yourself, then call [the *instance* `Attach<T>` method](https://docs.microsoft.com/en-us/dotnet/api/streamjsonrpc.jsonrpc.attach?view=streamjsonrpc-2.9#streamjsonrpc-jsonrpc-attach-1(streamjsonrpc-jsonrpcproxyoptions)) to create your proxy object, add your event handlers, then call `JsonRpc.StartListening`.

**Reply by csehatt741** [reply]
Unfortunately, correcting the property name and adding the attribute did not help.

```csharp
public class CustomEventArgs : EventArgs
{
    [JsonProperty("value")]
    public string Value { get; set; }
}
```

**Reply by csehatt741** [reply]
As `JsonRpcRequest.DebuggerDisplay` is `Notification: onEvent(System.Collections.Generic.Dictionary``2[System.String,System.Object])`, so I also tried out this:

``` csharp
public class CustomEventArgs : EventArgs
{
    [JsonProperty("value")]
    public Dictionary<string, object> Value { get; set; }
}
```

Which did not work either.

**Reply by AArnott** [reply]
When I suggested that your `value` arg name didn't match, I was talking about the *parameter*, not the property. In JSON-RPC terms, the incoming message you quoted is using named argument syntax, and the only argument is named `value`. But your event handling lambda calls the parameter `args`. So I'm suggesting you try renaming the parameter in your lambda to `value`.
